Fix read boundary

moxcodes opened the pull request:
## Proposed changes

fix #1950 
fix #1949 

### Types of changes:

- [x] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

kidder commented:
okay, the second commit was about what I expected; the first was not... and it apparently doesn't work????

moxcodes commented:
Yeah, sorry, should have checked the conversion script executable as well.
The first commit has a bunch of changes because I actually ran into problems that the test I wrote would have caught, but I failed to call the test function in the test case in the original PR.

kidder commented:
Can you please list the things that were broken to make it easier to review the changes?

moxcodes commented:
Sure, I am also happy to split any subset of these you'd like into separate commits:
- The boundary data routines need to compute both du r / r and du r because one version of the input pipeline uses one, and the other uses the other. They're easy enough to infer from one another, so the cost is low to just generalize and use both always.
- the buffer updater needs to know the l_max associated with the computation, which might be smaller or larger than the file l_max, so that it can either produce a padded set of modes or omit modes when it gives the data to the interpolator.
- For the reduced file type, the buffer updater needs to look at a different dataset to initialize the list of times
- the buffer updater needs to adjust behavior depending on whether or not the quantity is real (spin 0) and therefore has half the modes omitted in the reduced version of the data set.
- the pup function needs to actually pack the buffer updater because otherwise nothing works when the component is migrated
- I've made the `populate_hypersurface_boundary_data` member functions const, which will be needed for easy use.
- the reduced data set gives du_j, not h, so the typelists need to be updated accordingly

kidder commented:
yes, please, split the bullets into separate commits if you don't mind; they can be squashed together after review...

kidder commented:
and maybe split off what was the file rename commit into a separate PR so we can merge it

moxcodes commented:
Opened #1953

moxcodes committed:
Fix ReadBoundaryH5 utility, enable ignored test

- The boundary data routines need to compute both du r / r and du r because one
  version of the input pipeline uses one, and the other uses the other.
  They're easy enough to infer from one another, so the cost is low to just
  generalize and use both always.
- the buffer updater needs to know the l_max associated with the computation,
  which might be smaller or larger than the file l_max, so that it can either
  produce a padded set of modes or omit modes when it gives the data to the
  interpolator.
- For the reduced file type, the buffer updater needs to look at a different
  dataset to initialize the list of times
- the buffer updater needs to adjust behavior depending on whether or not the
  quantity is real (spin 0) and therefore has half the modes omitted in the
  reduced version of the data set.
- the pup function needs to actually pack the buffer updater because otherwise
  nothing works when the component is migrated
- I've made the populate_hypersurface_boundary_data member functions const,
  which will be needed for easy use.
- the reduced data set gives du_j, not h, so the typelists need to be updated
  accordingly

moxcodes commented:
broken up into smaller commits.
Cheers!

nilsvu commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.cpp:
```diff
@@ -262,7 +262,8 @@ ReducedSpecWorldtubeH5BufferUpdater::ReducedSpecWorldtubeH5BufferUpdater(
   get<Tags::detail::InputDataSet<
-      Spectral::Swsh::Tags::SwshTransform<Tags::BondiH>>>(dataset_names_) = "H";
+      Spectral::Swsh::Tags::SwshTransform<Tags::Du<Tags::BondiJ>>>>(
+      dataset_names_) = "H";
```
should the label also change?

moxcodes replied:
It shouldn't, but the reasoning is a little subtle: because H is coordinate dependent, so what's in the file is H in bondi coordinates (which makes sense for what is stored), but what we call H in the code is the H in numerical coordinates, because that is what is needed for the time stepping.

nilsvu commented:
I would have found it useful if you had added the bullet points above to each commit's message. Could still be worth doing for the next reviewer :)

kidder reviewed (commented):
with @nilsdeppe go ahead and squash into a single commit, but put your bullet list in the commit message

moxcodes commented:
squashed an commit message updated.
Cheers!
